Loan Process - 2nd version

salony75way opened the pull request:
fixes #232

IanMayo commented:
Compile error to be fixed, please @salony75way :

Error: src/providers/dataProvider/index.ts(83,11): error TS2339: Property 'name' does not exist on type 'true | { name: string; }'.

IanMayo commented:
Hello @salony75way - the architecture still isn't quite right. We no longer require the `Loan` entity.  We just have some loan-related attributes to `Item`.   Updated schema [here](https://docs.google.com/drawings/d/1SYIb0PZxcabvhXdNMAEHlbdWaoam2ipB7x5ZUudVVGc/edit)

salony75way commented:
> Hello @salony75way - the architecture still isn't quite right. We no longer require the `Loan` entity. We just have some loan-related attributes to `Item`. Updated schema [here](https://docs.google.com/drawings/d/1SYIb0PZxcabvhXdNMAEHlbdWaoam2ipB7x5ZUudVVGc/edit)

Hi @IanMayo , Yes i have removed all resources related to loan. Let me know if we need to remove some other dependency in code.

IanMayo commented:
> Hi @IanMayo , Yes i have removed all resources related to loan. 

We still have the `Loan` type, and we still refer to it from `LoanForm`:
![image](https://user-images.githubusercontent.com/1108513/236423566-245e9c29-7446-4c50-bb08-ef9b69f1c3e9.png)

IanMayo commented:
One other change please. It's about `loanedBy`.  This suggests it is the librarian that lent the book. This should be `loanedTo` - so it's the person that has taken the book away.

IanMayo commented:
Hello @salony75way - IMHO the `loan` form and buttons should be moved to the `Items` resource - since they are only used for that resource.

IanMayo commented:
Aah, one more thing. Can we have a `notify` popup once the database changes are complete?

IanMayo commented on src/providers/dataProvider/index.ts:
```diff
@@ -78,33 +78,76 @@ const withCreatedBy = (
+        await audit({
+          type: AuditType.ITEM_LOAN,
+          activityDetail: `Item loaned to ${name} by ${userName}. ${remarks}`,
+          resource: constants.R_LOAN_ITEMS,
```
This should be `constants.R_ITEMS`

We should probably delete `constants.R_LOAN_ITEMS`

IanMayo commented on src/types.d.ts:
```diff
@@ -108,6 +103,7 @@ interface Item {
   musterRemarks: string
   createdAt: string
   createdBy: User['id']
+  loanedBy?: number
```
this should be `loanedTo`

IanMayo commented:
Have just been talking to users. They have said they don't need the `Remarks` field in the `LoanForm`.   It's not on the legacy system, nor is it in the system their sister-organisation uses.

So, it will get deleted from the `remarks` too, please.
